pkg/gardenlet/controller/extensions/seed_control.go

```diff
@@ -0,0 +1,236 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensions
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+	gardencorelisters "github.com/gardener/gardener/pkg/client/core/listers/core/v1alpha1"
+	gardenextensionsclient "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	gardenextensionsinformers "github.com/gardener/gardener/pkg/client/extensions/informers/externalversions"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+	"github.com/gardener/gardener/pkg/logger"
+	seedpkg "github.com/gardener/gardener/pkg/operation/seed"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// FederatedSeedController is responsible for maintaining multiple federated Seeds' controllers
+type FederatedSeedController struct {
+	federatedSeedControllerManager *federatedSeedControllerManager
+	k8sGardenClient                kubernetes.Interface
+
+	config *config.GardenletConfiguration
+
+	seedLister    gardencorelisters.SeedLister
+	seedHasSynced cache.InformerSynced
+	seedQueue     workqueue.RateLimitingInterface
+
+	recorder record.EventRecorder
+
+	ctx                    context.Context
+	numberOfRunningWorkers int
+	workerCh               chan int
+}
+
+func (s *FederatedSeedController) seedAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+
+		return
+	}
+	s.seedQueue.Add(key)
+}
+
+func (s *FederatedSeedController) seedDelete(obj interface{}) {
+	seed := obj.(*gardencorev1alpha1.Seed)
+	s.federatedSeedControllerManager.removeController(seed.GetName())
+}
+
+func (s *FederatedSeedController) seedUpdate(oldObj, newObj interface{}) {
+	var (
+		newSeedObj = newObj.(*gardencorev1alpha1.Seed)
+		oldSeedObj = oldObj.(*gardencorev1alpha1.Seed)
+	)
+
+	newSeedBootstrappedCondition := helper.GetCondition(newSeedObj.Status.Conditions, gardencorev1alpha1.SeedBootstrapped)
+	oldSeedBootstrappedCondition := helper.GetCondition(oldSeedObj.Status.Conditions, gardencorev1alpha1.SeedBootstrapped)
+
+	if oldSeedBootstrappedCondition.Status != newSeedBootstrappedCondition.Status {
+		if newSeedBootstrappedCondition.Status == gardencorev1alpha1.ConditionTrue {
+			s.seedAdd(newSeedObj)
+		}
+		s.seedDelete(newSeedObj)
```

Comment:
Why do we remove the seed in this case? The seed should only be removed if the deletion timestamp is set, or?

Reply:
We don't have to remove it. We would only enter this case if the seed has already been bootstrapped meaning that the extensions have already been deployed.

Reply:
It was meant to stop the seed controller in case of the seed's condition - bootstrapped changes to False... but it seems that I have missed the return/else statement.